File: rasmview/src/main/java/com/raed/rasmview/touch/gesture/rotation/RotationGestureDetector.java
package com.raed.rasmview.touch.gesture.rotation;

import static java.lang.Math.abs;
import static java.lang.Math.atan;
import static java.lang.Math.toDegrees;

import android.view.MotionEvent;

public class RotationGestureDetector {

    private OnRotationListener mListener;

    public RotationGestureDetector(OnRotationListener listener) {
        mListener = listener;
    }

    public static float angleBetweenLines(float[][] line1, float[][] line2) {

        double denominator = (line1[1][0] - line1[0][0]);
        denominator = denominator == 0 ? 0.000000001 : denominator;
        double m1 = (line1[1][1] - line1[0][1])/denominator;

        denominator = (line2[1][0] - line2[0][0]);
        denominator = denominator == 0 ? 0.000000001 : denominator;
        double m2 = (line2[1][1] - line2[0][1])/denominator;

        if (m1 * m2 == -1)
            return 90;

        return  (m2 > m1 ? -1 : 1)  * (float) toDegrees(atan(abs((m1 - m2)/(1 + m1 * m2))));
    }

    private float[][] mCurrentLine = new float[2][2];
    private float[][] mLastLine = new float[2][2];
    private int m1stPointerID;
    private int m2ndPointerID;
    public void onTouchEvent(MotionEvent event) {
        if (event.getPointerCount() != 2)
            return;
        switch (event.getActionMasked()) {
            case MotionEvent.ACTION_POINTER_DOWN:
                m1stPointerID = event.getPointerId(0);
                m2ndPointerID = event.getPointerId(1);
                mLastLine[0][0] = event.getX(0);
                mLastLine[0][1] = event.getY(0);
                mLastLine[1][0] = event.getX(1);
                mLastLine[1][1] = event.getY(1);
                break;
            case MotionEvent.ACTION_MOVE:
            case MotionEvent.ACTION_UP:
                if (m1stPointerID != event.getPointerId(0) || m2ndPointerID != event.getPointerId(1))
                    return;
                mCurrentLine[0][0] = event.getX(0);
                mCurrentLine[0][1] = event.getY(0);
                mCurrentLine[1][0] = event.getX(1);
                mCurrentLine[1][1] = event.getY(1);
                mListener.onRotation(angleBetweenLines(mLastLine, mCurrentLine));
                mLastLine[0][0] = mCurrentLine[0][0];
                mLastLine[0][1] = mCurrentLine[0][1];
                mLastLine[1][0] = mCurrentLine[1][0];
                mLastLine[1][1] = mCurrentLine[1][1];
                break;
        }
    }

}

File: rasmview/src/main/java/com/raed/rasmview/touch/gesture/RotationGestureDetector.kt
package com.raed.rasmview.touch.gesture

import android.view.MotionEvent
import kotlin.math.abs
import kotlin.math.atan

class RotationGestureDetector(
    private val listener: (Float) -> Unit,
) {

    private val currentLine = Array(2) { FloatArray(2) }
    private val lastLine = Array(2) { FloatArray(2) }
    private var firstPointerID = 0
    private var secondPointerID = 0

    fun onTouchEvent(event: MotionEvent) {
        if (event.pointerCount != 2) return
        when (event.actionMasked) {
            MotionEvent.ACTION_POINTER_DOWN -> {
                firstPointerID = event.getPointerId(0)
                secondPointerID = event.getPointerId(1)
                lastLine[0][0] = event.getX(0)
                lastLine[0][1] = event.getY(0)
                lastLine[1][0] = event.getX(1)
                lastLine[1][1] = event.getY(1)
            }
            MotionEvent.ACTION_MOVE, MotionEvent.ACTION_UP -> {
                if (firstPointerID != event.getPointerId(0) || secondPointerID != event.getPointerId(1)) return
                currentLine[0][0] = event.getX(0)
                currentLine[0][1] = event.getY(0)
                currentLine[1][0] = event.getX(1)
                currentLine[1][1] = event.getY(1)
                listener(angleBetweenLines(lastLine, currentLine))
                lastLine[0][0] = currentLine[0][0]
                lastLine[0][1] = currentLine[0][1]
                lastLine[1][0] = currentLine[1][0]
                lastLine[1][1] = currentLine[1][1]
            }
        }
    }

}

private fun angleBetweenLines(line1: Array<FloatArray>, line2: Array<FloatArray>): Float {
    var denominator = (line1[1][0] - line1[0][0]).toDouble()
    denominator = if (denominator == 0.0) 0.000000001 else denominator
    val m1 = (line1[1][1] - line1[0][1]) / denominator
    denominator = (line2[1][0] - line2[0][0]).toDouble()
    denominator = if (denominator == 0.0) 0.000000001 else denominator
    val m2 = (line2[1][1] - line2[0][1]) / denominator
    return if (m1 * m2 == -1.0) {
        90f
    } else {
        (if (m2 > m1) -1 else 1) * Math.toDegrees(atan(abs((m1 - m2) / (1 + m1 * m2)))).toFloat()
    }
}